platform/FileHandle.h

```diff
@@ -152,8 +162,106 @@ class FileHandle {
      */
     MBED_DEPRECATED_SINCE("mbed-os-5.4", "Replaced by FileHandle::size")
     virtual off_t flen() { return size(); }
+
+    /** Set blocking or non-blocking mode of the file operation like read/write.
+     *  Definition depends upon the subclass implementing FileHandle.
+     *  The default is blocking.
+     *
+     *  @param blocking true for blocking mode, false for non-blocking mode.
+     */
+    virtual int set_blocking(bool blocking) {
+        return -1;
+    }
+
+    /** Check for poll event flags
+     * The input parameter can be used or ignored - the could always return all events,
+     * or could check just the events listed in events.
+     * Call is non-blocking - returns instantaneous state of events.
+     * Whenever an event occurs, the derived class must call _poll_change().
+     * @param events bitmask of poll events we're interested in - POLLIN/POLLOUT etc.
+     * @return bitmask of poll events that have occurred.
+     */
+    virtual short poll(short events) const {
```

Comment:
My current thought on lack of true poll for some network stacks is that we make it report with an honest EINVAL if it doesn't work - fully-portable network applications considering using poll() would have to cope with and expect the EINVAL it would return, and have a backup plan (until the day they can rely on it being available on all stacks they care about). There would be a method in FileHandle to indicate the unavailablility.

POSIX mentions in passing returning EINVAL for cases of an fd being inappropriate for poll (as opposed to the POLLNVAL flag for fd not being valid at all).

There's a limit to what you can practically do with uncertainty - a definite poll lets you "safely" make blocking calls (although there are race pitfalls), but that's out if you're not certain. So you'd have to go for non-blocking calls. And the problem is that a timed/blocking poll won't work - the EINVAL would be immediate. You'd have to effectively just be in SIGIO + nonblocking operation. Maybe using poll() as a hint - you could try a non-blocking read if POLLIN report or EINVAL?

Maybe more intricacy could be added to do "better" for poll but it feels like encouraging the wrong behaviour. The most straightforward thing in that situation is to use SIGIO and non-blocking and not bother trying to work around a (potentially) non-functioning poll(). If we think a correctly-working SIGIO is too difficult for app authors, I don't see how we expect them to make sense of an occasionally-working poll().